Library/Homebrew/cask/lib/hbc/installer.rb

```diff
@@ -420,19 +434,43 @@ def zap
       purge_caskroom_path
     end
 
+    def backup_path
+      return nil if @cask.staged_path.nil?
+      Pathname.new "#{@cask.staged_path}.upgrading"
+    end
+
+    def version_is_latest?
+      @cask.versions.include?("latest")
+    end
+
     def gain_permissions_remove(path)
       Utils.gain_permissions_remove(path, command: @command)
     end
 
+    def purge_backed_versioned_files
+      ohai "Purging files for version #{@cask.version} of Cask #{@cask}"
+
+      # versioned staged distribution
+      gain_permissions_remove(backup_path) if !backup_path.nil? && backup_path.exist?
+
+      # Homebrew-Cask metadata
+      purge_metadata
+    end
+
     def purge_versioned_files
       ohai "Purging files for version #{@cask.version} of Cask #{@cask}"
 
       # versioned staged distribution
       gain_permissions_remove(@cask.staged_path) if !@cask.staged_path.nil? && @cask.staged_path.exist?
 
       # Homebrew-Cask metadata
+      purge_metadata
+    end
+
+    def purge_metadata
```

Comment:
The lines in here which have `unless upgrade?` could be moved to `purge_versioned_files` now.